shen-sbcl: 3.0.3 -> 39.1

hakujin opened the pull request:
Migrates to the new "S"-style Shen kernels which brings with them a new build system. Removes the (now deprecated) shen-sources package.


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [X] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

donovanglover commented:
@bsima As the author of the original PR, are you still interested in maintaining this package? I can't request you as a reviewer since you aren't in the NixOS organization.

donovanglover reviewed (commented):
Please create multiple commits to follow [contributing guidelines](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions). Make sure that the maintainer commit comes before adding yourself as a maintainer of the package.

- maintainers: add hakujin
- shen-sbcl: 3.0.3 -> 39.1
- shen-sbcl: add hakujin as maintainer

donovanglover commented on pkgs/by-name/sh/shen-sbcl/package.nix:
```diff
@@ -1,43 +1,36 @@
-  preBuild = ''
-    ln -s ${shen-sources} kernel
+  src = fetchzip {
+    url = "https://www.shenlanguage.org/Download/${kernel-version}.zip";
```
See https://github.com/NixOS/nixpkgs/pull/336221/commits/41fb6f1f0d988249d0e0c046781d4fc49b6cd13c for an example of how to implement the `finalAttrs` pattern

hakujin replied:
> Are these sources available in a git repository?

Sources are only officially available from the language's website.

donovanglover replied:
```suggestion
    url = "https://www.shenlanguage.org/Download/${lib.replaceStrings [ "0-" ] [ "" ] finalAttrs.version}.zip";
```
`finalAttrs` is preferred over `let-in` so `version` can be overwritten. See https://github.com/NixOS/nixpkgs/pull/387439#issuecomment-2702724085

hakujin replied:
Done.

Commit:
maintainers: add hakujin

Adds hakujin (Colin King) to maintainers.

Commit:
shen-sbcl: 3.0.3 -> 39.1

Migrates to the new "S"-style Shen kernels which brings with them a new
build system. Removes the (now deprecated) shen-sources package.

Commit:
shen-sbcl: add hakujin as maintainer

Adds hakujin (Colin King) as a maintainer for `shen-sbcl`.

hakujin commented on pkgs/by-name/sh/shen-sbcl/package.nix:
```diff
@@ -1,43 +1,36 @@
-    sha256 = "0mc10jlrxqi337m6ngwbr547zi4qgk69g1flz5dsddjy5x41j0yz";
-  };
-
+  version = "0-${kernel-version}";
```
Some context: the "S" part of the version is meaningful; previously Shen used plain numeric versions and the migration to the new S kernels brought a lot of changes, both semantically for the language and also for the build system. I think having the "S" as part of the version is important.

If you still think this should be changed let me know.

donovanglover replied:
This would be the first package in nixpkgs that follows this format based on `rg "version = \"0-[A-Za-tv-z]"` so it could break some things. The [guideline](https://github.com/NixOS/nixpkgs/blame/7a5c9e3fa99025c13f833cbd70ce0ed933f4ea05/pkgs/README.md#L356) was added in https://github.com/NixOS/nixpkgs/pull/280312 however the interpretation here doesn't seem to be used in practice; version numbers or `0-unstable-YYYY-MM-DD` are preferred wherever possible.

hakujin commented on pkgs/by-name/sh/shen-sbcl/package.nix:
```diff
@@ -1,43 +1,36 @@
   installPhase = ''
-    install -m755 -D bin/sbcl/shen $out/bin/shen-sbcl
+    mkdir -p $out/bin
+    cp sbcl-shen.exe $out/bin/shen-sbcl
```
Done, thanks!

hakujin commented on pkgs/by-name/sh/shen-sbcl/package.nix:
```diff
@@ -1,43 +1,36 @@
+let
+  kernel-version = "S39.1";
+in
+stdenv.mkDerivation {
```
Done, thanks!

hakujin commented on pkgs/by-name/sh/shen-sbcl/package.nix:
```diff
@@ -1,43 +1,36 @@
+    sha256 = "sha256-reN9avgYGYCMiA5BeHLhRK51liKF2ctqIgxf+4IWjVY=";
+  };
+  dontStrip = true; # necessary to prevent runtime errors with sbcl
+  dontConfigure = true;
```
Done, thanks!

hakujin commented on pkgs/by-name/sh/shen-sbcl/package.nix:
```diff
@@ -1,43 +1,36 @@
+  dontStrip = true; # necessary to prevent runtime errors with sbcl
+  dontConfigure = true;
+  buildPhase = ''
+    sbcl --noinform --no-sysinit --no-userinit --load install.lsp
```
Done, thanks!

donovanglover commented:
Pending CI checks are common and not a blocker for merges. Tools like [`nixpkgs-review-gha`](https://github.com/Defelo/nixpkgs-review-gha) are often used to confirm that packages build successfully.

We still have to review other parts of the PR, which can take a while depending on how many people with the adequate knowledge are interested in it.

donovanglover reviewed (commented):
LGTM after changing the commit message to `shen-sbcl: 3.0.3 -> 39.1`. Also left some nits to align the package with conventions

hakujin commented:
> LGTM after changing the commit message to `shen-sbcl: 3.0.3 -> 39.1`. Also left some nits to align the package with conventions

Thanks for the feedback; I believe everything has been addressed and this is ready to go.

donovanglover commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review-gha`](https://github.com/Defelo/nixpkgs-review-gha)

Command: `nixpkgs-review pr 380807`

Logs: https://github.com/donovanglover/nixpkgs-review-gha/actions/runs/13826754550


---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>shen-sbcl</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>shen-sbcl</li>
  </ul>
</details>

---
### `x86_64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>shen-sbcl</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>shen-sbcl</li>
  </ul>
</details>

donovanglover reviewed (approved):
Tested and LGTM grats on your first contribution to nixpkgs!

Edit: Actually this isn't your first contribution but welcome to the maintainers list!
